py3status/modules/playerctl.py

```diff
@@ -0,0 +1,272 @@
+        seconds = microseconds // 1_000_000
+        m, s = divmod(seconds, 60)
+        h, m = divmod(m, 60)
+        time = f"{h}:{m:02d}:{s:02d}"
+        return time.lstrip("0").lstrip(":")
+
+    def _get_player_position(self, player):
+        try:
+            # Playerctl doesn't support getting positions for all players
+            microseconds = player.get_position()
+            position = self._microseconds_to_time(microseconds)
+        except GLib.GError:
+            position = None
+
+        return position
+
+    def _set_data_from_metadata(self, player, data):
+        """Set any data retrieved directly from the metadata for a player"""
+        metadata = dict(player.props.metadata)
+
+        duration_ms = metadata.get("mpris:length")
+        if duration_ms:
+            data["duration"] = self._microseconds_to_time(duration_ms)
+        else:
+            data["duration"] = None
+
+    def _get_player_data(self, player):
+        data = {}
+
+        # Song attributes
+        data["album"] = player.get_album()
+        data["artist"] = player.get_artist()
+        data["title"] = player.get_title()
+        data["position"] = self._get_player_position(player)
+        self._set_data_from_metadata(player, data)
+
+        # Player attributes
+        data["player"] = player.props.player_name
+        data["status"] = player.props.status.lower()
+
+        return data
+
+    def playerctl(self):
+        tracked_players = self.manager.props.players
+
+        # Only have a cache timeout if the user wants to know the players' positions
+        if tracked_players and (
+            self.py3.format_contains(self.format, "position")
+            or self.py3.format_contains(self.format_player, "position")
+        ):
+            cached_until = 1
+        else:
+            cached_until = self.py3.CACHE_FOREVER
+
+        players = []
+        for player in tracked_players:
+            player_data = self._get_player_data(player)
+
+            # Set the color of a player
+            for key in self.thresholds_init:
+                if key in player_data:
+                    self.py3.threshold_get_color(player_data[key], key)
+
+            format_player = self.py3.safe_format(self.format_player, player_data)
+            self.py3.composite_update(format_player, {"index": player_data["player"]})
+
+            players.append(format_player)
+
+        format_player_separator = self.py3.safe_format(self.format_player_separator)
+        format_players = self.py3.composite_join(format_player_separator, players)
+
+        return {
+            "cached_until": self.py3.time_in(cached_until),
+            "full_text": self.py3.safe_format(
+                self.format, {"format_player": format_players}
+            ),
+        }
+
+    def _get_player_from_index(self, index):
+        for player in self.manager.props.players:
+            if player.props.player_name == index:
+                return player
+
+        return None
```

Comment:
(Nitpick) Please move this func up. I care that last two funcs are for py3status's... and other `underscore_funcs` stays grouped on top. _....AKA I like to keep them separate._